src/mscorlib/corefx/System/Buffers/MultiProducerMultiConsumerBoundedQueue.cs

```diff
@@ -0,0 +1,172 @@
+        /// This must be greater than 1 and must be a power of 2.
+        /// </summary>
+        public MultiProducerMultiConsumerBoundedQueue(int boundedLength)
+        {
+            Debug.Assert(boundedLength >= 2, $"Must be >= 2, got {boundedLength}");
+            Debug.Assert((boundedLength & (boundedLength - 1)) == 0, $"Must be a power of 2, got {boundedLength}");
+
+            _slots = new Slot[boundedLength];
+            _slotsMask = boundedLength - 1;
+
+            // Initialize the sequence number for each slot
+            for (int i = 0; i < _slots.Length; i++)
+            {
+                _slots[i].SequenceNumber = i;
+            }
+        }
+
+        /// <summary>Dequeues an element from the queue. Null is returned if no element was available.</summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public T TryDequeue()
+        {
+            // Loop in case of contention...
+            var spinner = new SpinWait();
+            while (true)
+            {
+                // Get the head at which to try to rent.
+                long currentHead = Volatile.Read(ref _headAndTail.Head);
+                int slotsIndex = (int)(currentHead & _slotsMask);
+
+                // Read the sequence number for the head position.
+                long sequenceNumber = Volatile.Read(ref _slots[slotsIndex].SequenceNumber);
+
+                // Sequence numbers are targeted at returns, and are one higher for rents.
+                // So compare the sequence number to head+1.  If they're the same, then this is our
+                // element to dequeue, but we need to use an interlocked to advance the head, as
+                // concurrent renters could take it out from under us. If the sequence number is less
+                // than head+1, then there's nothing to rent, so we return null.  If the sequence
+                // number is greater than head+1, we lost a race with some other thread that incremented
+                // the sequence number, so we need to try again.
+                long delta = sequenceNumber - (currentHead + 1);
+                if (delta == 0)
+                {
+                    if (Interlocked.CompareExchange(ref _headAndTail.Head, currentHead + 1, currentHead) == currentHead)
+                    {
+                        // Successfully reserved the slot.  Note that after the above CompareExchange, other threads
+                        // trying to rent will end up spinning until we do the subsequent Write.
+                        T item = _slots[slotsIndex].Item;
+                        _slots[slotsIndex].Item = null;
+                        Volatile.Write(ref _slots[slotsIndex].SequenceNumber, currentHead + _slots.Length);
+                        return item;
+                    }
+                }
+                else if (delta < 0)
+                {
+                    // Queue is empty
```

Comment:
Oh ok, missed the interlocked read on 32-bit platforms. It would be good to add a comment or to reflect in the name that enqueues/dequeues are not guaranteed, as the class could be used later for another purpose where it matters.